Mathlib/CategoryTheory/Topos/Classifier.lean

```diff
@@ -48,68 +49,126 @@ Let `C` refer to a category with a terminal object.
 
 * [S. MacLane and I. Moerdijk, *Sheaves in Geometry and Logic*][MM92]
 
-## TODO
-
-* Make API for constructing a subobject classifier without reference to limits (replacing `⊤_ C`
-  with an arbitrary `Ω₀ : C` and including the assumption `mono truth`)
-
 -/
 
 universe u v u₀ v₀
 
 namespace CategoryTheory
 
-open Category Limits Functor
+open Category Limits Functor IsPullback
 
-variable (C : Type u) [Category.{v} C] [HasTerminal C]
+variable (C : Type u) [Category.{v} C]
 
-/-- A morphism `truth : ⊤_ C ⟶ Ω` from the terminal object of a category `C`
-is a subobject classifier if, for every monomorphism `m : U ⟶ X` in `C`,
-there is a unique map `χ : X ⟶ Ω` such that the following square is a pullback square:
+/-- A monomorphism `truth : Ω₀ ⟶ Ω` is a subobject classifier if, for every monomorphism
+`m : U ⟶ X` in `C`, there is a unique map `χ : X ⟶ Ω` such that for some (necessarily unique)
+`χ₀ : U ⟶ Ω₀` the following square is a pullback square:
 ```
       U ---------m----------> X
       |                       |
-terminal.from U               χ
+      χ₀                      χ
       |                       |
       v                       v
-    ⊤_ C ------truth--------> Ω
+      Ω₀ ------truth--------> Ω
 ```
-An equivalent formulation replaces the object `⊤_ C` with an arbitrary object, and instead
-includes the assumption that `truth` is a monomorphism.
+An equivalent formulation replaces `Ω₀` with the terminal object.
 -/
 structure Classifier where
-  /-- The target of the truth morphism -/
+  /-- The domain of the truth morphism -/
+  Ω₀ : C
+  /-- The codomain of the truth morphism -/
   Ω : C
-  /-- the truth morphism for a subobject classifier -/
-  truth : ⊤_ C ⟶ Ω
+  /-- The truth morphism of the subobject classifier -/
+  truth : Ω₀ ⟶ Ω
+  /-- The truth morphism is a monomorphism -/
+  mono_truth : Mono truth
+  /-- The top arrow in the pullback square -/
+  χ₀ {U X : C} (m : U ⟶ X) [Mono m] : U ⟶ Ω₀
   /-- For any monomorphism `U ⟶ X`, there is an associated characteristic map `X ⟶ Ω`. -/
   χ {U X : C} (m : U ⟶ X) [Mono m] : X ⟶ Ω
-  /-- `χ m` forms the appropriate pullback square. -/
-  isPullback {U X : C} (m : U ⟶ X) [Mono m] : IsPullback m (terminal.from U) (χ m) truth
-  /-- `χ m` is the only map `X ⟶ Ω` which forms the appropriate pullback square. -/
-  uniq {U X : C} (m : U ⟶ X) [Mono m] (χ' : X ⟶ Ω)
-      (hχ' : IsPullback m (terminal.from U) χ' truth) :
-    χ' = χ m
+  /-- `χ m` and `χ₀ m` form the appropriate pullback square. -/
+  isPullback' {U X : C} (m : U ⟶ X) [Mono m] : IsPullback m (χ₀ m) (χ m) truth
+  /-- `χ m` is the only map `X ⟶ Ω` which forms the appropriate pullback square for any `χ₀'`. -/
+  uniq' {U X : C} (m : U ⟶ X) [Mono m] (χ₀' : U ⟶ Ω₀) (χ' : X ⟶ Ω)
+    (hχ' : IsPullback m χ₀' χ' truth) : χ' = χ m
+
+variable {C}
+
+namespace Classifier
+
+attribute [instance] mono_truth
+
+/-- More explicit constructor in case `Ω₀` is already known to be a terminal object. -/
```

Comment:
```suggestion
/-- More explicit constructor in case `Ω₀` is already known to be a terminal object. -/
@[simps]
```

You probably want `@[simps]` to avoid unfolding every field of a classifier constructed this way in practice.